src/DataSource.cc

```diff
@@ -0,0 +1,188 @@
+#include "podio/DataSource.h"
+#include "podio/Reader.h"
+
+// podio
+#include <podio/FrameCategories.h>
+
+// ROOT
+#include <TFile.h>
+
+// STL
+#include <cstddef>
+#include <cstdio>
+#include <memory>
+
+namespace podio {
+DataSource::DataSource(const std::string& filePath, int nEvents) : m_nSlots{1} {
+  m_filePathList.emplace_back(filePath);
+  SetupInput(nEvents);
+}
+
+DataSource::DataSource(const std::vector<std::string>& filePathList, int nEvents) :
+    m_nSlots{1}, m_filePathList{filePathList} {
+  SetupInput(nEvents);
+}
+
+void DataSource::SetupInput(int nEvents) {
+  if (m_filePathList.empty()) {
+    throw std::runtime_error("podio::DataSource: No input files provided!");
+  }
+
+  // Check if the provided file(s) exists and contain required metadata is done
+  // by podio::Reader
+
+  // Create probing frame
+  podio::Frame frame;
+  unsigned int nEventsInFiles = 0;
+  auto podioReader = podio::makeReader(m_filePathList);
+  nEventsInFiles = podioReader.getEntries(podio::Category::Event);
+  frame = podio::Frame(podioReader.readFrame(podio::Category::Event, 0));
+
+  // Determine over how many events to run
+  if (nEventsInFiles > 0) {
+  } else {
+    throw std::runtime_error("podio::DataSource: No events found!");
+  }
+
+  if (nEvents < 0) {
+    m_nEvents = nEventsInFiles;
+  } else if (nEvents == 0) {
+    throw std::runtime_error("podio::DataSource: Requested to run over zero events!");
+  } else {
+    m_nEvents = nEvents;
+  }
+  if (nEventsInFiles < m_nEvents) {
+    m_nEvents = nEventsInFiles;
+  }
+
+  // Get collections stored in the files
+  std::vector<std::string> collNames = frame.getAvailableCollections();
+  for (const auto& collName : collNames) {
+    const podio::CollectionBase* coll = frame.get(collName);
+    if (coll->isValid()) {
+      m_columnNames.emplace_back(collName);
+      m_columnTypes.emplace_back(coll->getTypeName());
+    }
+  }
+}
+
+void DataSource::SetNSlots(unsigned int nSlots) {
+  m_nSlots = nSlots;
+
+  if (m_nSlots > m_nEvents) {
+    throw std::runtime_error("podio::DataSource: Number of events too small!");
+  }
+
+  int eventsPerSlot = m_nEvents / m_nSlots;
+  for (size_t i = 0; i < (m_nSlots - 1); ++i) {
+    m_rangesAll.emplace_back(eventsPerSlot * i, eventsPerSlot * (i + 1));
+  }
+  m_rangesAll.emplace_back(eventsPerSlot * (m_nSlots - 1), m_nEvents);
+  m_rangesAvailable = m_rangesAll;
+
+  // Initialize set of addresses needed
+  m_Collections.resize(m_columnNames.size(), std::vector<const podio::CollectionBase*>(m_nSlots, nullptr));
+
+  // Initialize podio readers
+  for (size_t i = 0; i < m_nSlots; ++i) {
+    m_podioReaders.emplace_back(std::make_unique<podio::Reader>(podio::makeReader(m_filePathList)));
+  }
+
+  for (size_t i = 0; i < m_nSlots; ++i) {
+    m_frames.emplace_back(std::make_unique<podio::Frame>(podio::Frame{}));
```

Comment:
```suggestion
    m_frames.emplace_back(std::make_unique<podio::Frame>());
```
make_unique forwards to c'tor, so there is no need to first create a frame and then make make_unique call the copy c'tor ?

Reply:
Thx, removed.